### ISSUE TITLE
Telemetry scripting framework API and docs

### ISSUE THREAD

**Issue by jsiwek** [zeek/zeek#1491]
Some follow-up tasks to the addition of C++/BIF Telemetry APIs in #1421:
* Add a script-layer framework API built on top of BIFs (has potential to more convenient, less verbose than BIF API in terms of typing)
* Add "framework" docs: overview of framework and available metrics, script & C++ examples, and Prometheus export how-to
* Add any missing Zeekygen comments/docs

**Comment by Neverlord** [comment]
Here's a starting point for the Telemetry documentation: https://github.com/zeek/zeek/wiki/Telemetry. So far it covers how to get Zeek to expose the Metrics via Prometheus and gives a brief overview plus some links to useful resources/examples (like the btest suite for the BIFs).

**Comment by rsmmr** [comment]
@grigorescu what parts are you working on here?

**Comment by grigorescu** [comment]
I've been working on the script-layer API to start. A couple of issues that I've run into have to do with how labels are stored.

For a family, the BIFs store labels as a string vector. This feels a bit unintuitive to me, and I'm wondering if it makes sense to use a `set[string]` instead. I think uniqueness matters more than order.

For a metric, the BIFs store labels as a `table[string] of string`, allowing things like `{["dim1"] = "vec1", ["dim2"] = "vec2"}`. The label key-value pairs are what define a specific metric. There is also a performance penalty associated with accessing a metric, so I was planning on building a cache along the lines of:

``` zeek
type LabelValues: table[string] of string;

# Some casts can't happen, so we just store the appropriate type.
type MetricOpaque: record {
	int_counter: opaque of int_counter_metric &optional;

	dbl_counter: opaque of dbl_counter_metric &optional;

	int_gauge: opaque of int_gauge_metric &optional;

	dbl_gauge: opaque of dbl_gauge_metric &optional;
};


cache: table[LabelValues] of MetricOpaque
```

The issue is that a table can't be the index value, though. One solution would be to flatten the label values into a string, and use that as the index, but that feels bad.

I was hoping to push something that at least ran; maybe I'll remove the caching for now and push out what I have as a first step.

**Comment by rsmmr** [comment]
Switching to a set of strings makes sense to me (cc: @Neverlord).

Yeah, I'd go without the caching at first to get something in place and then we can see how performance works out.

Would be great to get something into 4.1 still here (cc'ing @sethhall who has been thinking about this too).

**Comment by Neverlord** [comment]
> Switching to a set of strings makes sense to me (cc: @Neverlord).

No strong opinion either way. On the lower level (BIF), Zeek needs to wrap the strings into a contiguous container of type `vector<string_view>` though. If the scripts pass in a set, we'd need to provide a new conversion function (see https://github.com/zeek/zeek/blob/master/src/telemetry/telemetry.bif#L40).

> I was hoping to push something that at least ran; maybe I'll remove the caching for now and push out what I have as a first step.

Let me know if there's something on the 'lower levels' (BIF/C++) that doesn't quite fit or you need some assistance with. 🙂

**Comment by awelzel** [comment]
I'm closing this. It has been merged via #2273
